runtime/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -2933,46 +3181,20 @@ private void onKafkaData(
             {
                 doKafkaReset(traceId);
                 delegate.doMqttAbort(traceId, authorization);
-            }
-            else
-            {
-                handleKafkaData(data);
-            }
-        }
-
-        protected abstract void doKafkaBegin(
-            long traceId,
-            long authorization,
-            long affinity);
-
-        protected abstract void handleKafkaData(
-            DataFW data);
-
-        protected void onKafkaWindow(
-            WindowFW window)
-        {
-            final long sequence = window.sequence();
-            final long acknowledge = window.acknowledge();
-            final int maximum = window.maximum();
-
-            assert acknowledge <= sequence;
-            assert acknowledge >= delegate.initialAck;
-            assert maximum >= delegate.initialMax;
-
-            initialAck = acknowledge;
-            initialMax = maximum;
-            state = MqttKafkaState.openInitial(state);
-
-            assert initialAck <= initialSeq;
+            }
+            else
+            {
+                handleKafkaData(data);
+            }
```

Comment:
Check all `onKafkaFlush` to set `replySeq = sequence + reserved;` not just `replySeq = sequence;` which will handle the case when `reserved` is non-zero as well.

Suggest renaming to `onKafkaDataImpl(DataFW)` instead of `handleKafkaData`.